### ISSUE TITLE
assertion/deadlock when starting second `osrm-routed` consumer of shared memory block on OSX

### ISSUE THREAD

**Issue by danpat** [Project-OSRM/osrm-backend#3619]
When compiled with `-DCMAKE_BUILD_TYPE=Debug`, the following steps trigger an assertion from inside boost in one process, and a deadlock in another, on OSX:

```
#!/bin/sh

./osrm-datastore berlin.osrm
./osrm-routed -s &
./osrm-routed -s  # assertion here
```

The assertion that's thrown is:

```
$ ./osrm-routed -s
[info] starting up engines, v5.5.0
[info] Loading from shared memory
[info] Threads: 8
[info] IP address: 0.0.0.0
[info] IP port: 5000
[info] http 1.1 compression handled by zlib version 1.2.5
[info] Listening on: 0.0.0.0:5000
[debug] opening 76939265 from id 0
[debug] Loading new data for region REGION_2
[debug] opening 34471938 from id 2
[info] set checksum: 4127811523
[debug] opening 76939265 from id 0
[info] running and waiting for re[assert][q0x700000081000u] e/Users/danpat/mapbox/osrm-backend/mason_packages/headers/boost/1.63.0/include/boost/interprocess/sync/posix/condition.hpps:t175s
in: oid boost::interprocess::ipcdetail::posix_condition::do_wait(boost::interprocess::ipcdetail::posix_mutex &)[: 0res == 0m

libc++abi.dylib: terminating
Abort trap: 6
```

In addition, when attempting to kill the first `osrm-routed` process with CTRL+C, it fails to exit, hanging with this output:

```
^C[info] initiating shutdown
[info] stopping threads
[info] freeing objects
```

When run under `Release` mode, the assertion inside BOOST does not trigger, and everything looks normal (both `osrm-routed -s` processes continuously swap data and answer queries).

Note: the same behaviour is seen with `node-osrm` and instantiating two `OSRM()` objects.

**Comment by danpat** [comment]
Output from LLDB at the assertion point:

```
(lldb) n
Process 35936 stopped
* thread #2: tid = 0x1ae45b, 0x00000001002201cf osrm-routed`boost::interprocess::ipcdetail::posix_condition::do_wait(this=0x0000000101297050, mut=0x0000000101297010) + 47 at condition.hpp:175, stop reason = step over
    frame #0: 0x00000001002201cf osrm-routed`boost::interprocess::ipcdetail::posix_condition::do_wait(this=0x0000000101297050, mut=0x0000000101297010) + 47 at condition.hpp:175
   172 	   pthread_mutex_t* pmutex = &mut.m_mut;
   173 	   int res = 0;
   174 	   res = pthread_cond_wait(&m_condition, pmutex);
-> 175 	   BOOST_ASSERT(res == 0); (void)res;
   176 	}
   177
   178 	inline bool posix_condition::do_timed_wait
(lldb) p res
(int) $1 = 22
(lldb) p errno
(void *) $2 = 0x0000000000000011
```

(errno 0x11 is `File exists` from `strerror(0x11)`, I'm not sure if this is relevant).

Under OSX (from `/usr/include/sys/errno.h`), code 22 is

```
#define EINVAL          22              /* Invalid argument */
```

which is the same under Linux (I believe these are POSIX values).

The OSX manpage for `pthread_cond_wait` is a little different to Linux, it only describes a single error condition:

```
ERRORS
     pthread_cond_wait() will fail if:

     [EINVAL]           The value specified by cond or the value specified by
                        mutex is invalid.

```

**Comment by oxidase** [comment]
The problem is in undefined behavior in Boost.Interprocess and the POSIX interface and described [here](http://stackoverflow.com/a/35383885)

In POSIX interface for `pthread_cond_wait`:
http://pubs.opengroup.org/onlinepubs/007908775/xsh/pthread_cond_wait.html
> The effect of using more than one mutex for concurrent pthread_cond_wait() or pthread_cond_timedwait() operations on the same condition variable is undefined

`Boost.Interprocess` uses correctly shared between processes `pthread_cond_t` and `pthread_mutex_t` as declared [here](https://github.com/boostorg/interprocess/blob/develop/include/boost/interprocess/sync/shm/named_condition.hpp#L133-L145), but there is no guarantee that the shared memory block will be mapped into one address of different processes.

On the other side OSX  reports `EINVAL` on line 493  of https://opensource.apple.com/source/libpthread/libpthread-137.1.1/src/pthread_cond.c at `if (cond->busy != NULL && cond->busy != mutex)`  because it store on line 515 a user-space address of the mutex in the first process, so the check in the second process will compare a user-space address of the same mutex that will be almost always different.

One possibility would be to use fixed address mapping, but it would require change [here](https://github.com/boostorg/interprocess/blob/develop/include/boost/interprocess/sync/shm/named_condition.hpp#L188) from 0 to some address and not so obvious how to make it.

IMHO, it is not critical unless some one will need running multiple `osrm-routed` processes on OSX.

Will try to find another signaling mechanism or fix it for POSIX by creating a shared barrier based on `shared_memory_object` with fixed user space address. 

/cc @TheMarex, @danpat

**Comment by TheMarex** [comment]
This was fixed by #3627. Remaining issues captured in https://github.com/Project-OSRM/osrm-backend/issues/3633
